src/planning.rs

```diff
@@ -310,6 +311,20 @@ impl<'fetcher> BuildPlannerImpl<'fetcher> {
 
       let data_attr = possible_crate_settings.and_then(|s| s.data_attr.clone());
 
+      let git_data = if let Some(ref s) = own_package.source {
```

Comment:
Nit: This could be more idiomatically written something like this

```rust
let git_data = own_package.source
  .as_ref()
  .filter(SourceId::is_git)
  .map(|src| GitData {
    remote: src.url().into_string().to_owned(),
    // TODO: Explain why this unwrap is safe
    // UNRAP: e.g. like this
    commit: src.precise().unwrap().to_owned()
  })
```

Reply:
Looks like the internals use `SourceId#precise` in this manner:
https://github.com/rust-lang/cargo/blob/805fbebc830174ad34feb774e019769d6fe7bbe9/src/cargo/sources/git/source.rs#L33-L36
```rust
        let reference = match source_id.precise() {
            Some(s) => GitReference::Rev(s.to_string()),
            None => source_id.git_reference().unwrap().clone(),
        };
```

This has some kind of relationship to `GitReference`
https://github.com/rust-lang/cargo/blob/805fbebc830174ad34feb774e019769d6fe7bbe9/src/cargo/core/source/source_id.rs#L57-L64
```rust
pub enum GitReference {
    /// from a tag
    Tag(String),
    /// from the HEAD of a branch
    Branch(String),
    /// from a specific revision
    Rev(String),
}
```

---

That all said, we might not need to handle all of these edge cases? What does [cargo-vendor](https://github.com/alexcrichton/cargo-vendor) do?

Reply:
Not 100% sure on cargo-vendor, but I'd argue that we can only support Rev() or Tag() under Bazel. The `new_git_repository` rule doesn't like branches, and is against Bazel philosophy, for not being able to uniquely identify code state somewhere else.

Reply:
Final note - Option filtering is [nightly-only](https://github.com/rust-lang/rust/issues/45860) at the moment.